### PR TITLE
JavaArchipelago wrapper class for JavaProject visualization

### LINKED ISSUE #18: Island placement algorithm
Input: a `x,y` screen size, and clusters of nodes with sizes, and directed edges between them

Output: `x,y` coordinates representing the center of each node, such that edge lengths are relatively equal, and no edges cross (this supposedly looks aesthetically appealing)

### LINKED ISSUE #22: Create Typescript types for JavaProject and its parts

### LINKED ISSUE #38: Create relationships between Nodes (JavaClass; parts of Class) so we can have arrows
Convert JavaProject's parts into nodes and (weighted) directed edges. (This is required for #18, which will place the nodes based on the edges)

### PR DESCRIPTION
**Pull request by john-zou**
image

### `JavaArchipelago`
(Means group of islands, represents the entire Java Project, multiple packages)
- `width`, `height` for entire container
- `islands`: array of `JavaIsland`
- `islandMap`: an map of island name to `JavaIsland` instance, for efficient lookup of a `JavaIsland` by name. Every class can be looked up by its "fully qualified name"  (``` `${packageName}.${className}` ```). Only classes without a name conflict can be looked up by just its class name.
- `links`: representing relationships: contains all of the directed edges representing island-island dependencies.
  - `links.domesticDependencies` is for islands in the same package
  - `links.foreignDependencies` is for when two islands of two different packages have a dependency (represents imports). `source` is the one that's doing the export, and `target` is the one doing the import.

### `JavaIsland`
- extends `JavaFile`
- `center: { x, y }`
- `topLeftCorner: { x, y }`
- for convenience: `name`, `publicClass`, `otherClasses`, `package`

## Usage
Make a new `JavaArchipelago` passing in the `JavaProject` JSON, and then access its fields described above.

## To change their placement parameters (e.g. how close islands of a package are to each other):
Modify `util/constants.js`

---

Includes #23, #82 
Closes #18, closes #22, closes #38

### CODE REVIEW

File: client/src/JavaProjectTypes.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  constructors: JavaContructor[];
+};
+
+export type JavaPackage = {
+  name: string;
+  classes: JavaClass[];
+};
+
+export type JavaProject = {
+  packages: JavaPackage[];
```
Comment (MayKikuchi) [inline client/src/JavaProjectTypes.ts:55]: at least based on the sample output from Zhi before regarding the parsing, the value for packages seems to be an object

File: client/src/JavaProjectTypes.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
+export enum JavaType {
+  Interface = "Interface",
+  Enum = "Enum",
+  Abstract = "Abstract",
```
Comment (MayKikuchi) [inline client/src/JavaProjectTypes.ts:4]: I think it's currently "Abstract Class"

File: client/src/JavaProjectTypes.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  Concrete = "Concrete",
```
Comment (MayKikuchi) [inline client/src/JavaProjectTypes.ts:5]: Likewise, this is currently just "Class"

File: client/src/JavaProjectTypes.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  accessModifier: JavaAccessModifier;
+  parameters: JavaParameter[];
+};
+
+export type JavaClass = {
+  name: string;
+  type: JavaType;
+  accessModifier: JavaAccessModifier;
+  lineCount: number;
+  imports: string[];
```
Comment (MayKikuchi) [inline client/src/JavaProjectTypes.ts:43]: The imports are included in the file, and a file can have multiple classes

File: client/src/util/helpers.ts [hunk 5]
```diff
@@ -11,3 +13,21 @@ export function checkStringTypeIsCollection(type) {
...
 }
+
+export function getFullyQualifiedName(packageName: string, className: string) {
+  return `${packageName}.${className}`;
+}
+
+/**
+ * @returns not a whole number
+ */
+export function getApproximateIslandRadius(
```
[the remark this answers is not in the document]
Comment (MayKikuchi) [reply client/src/util/helpers.ts:24]: ^ updated comment

**Comment by algobot76** [comment]
Is there a way to render `JavaArchipelago` in react?

File: client/src/JavaProjectTypes.ts (hunk 1 above)
Comment (ZhiHanL) [reply client/src/JavaProjectTypes.ts:55]: Yup, can confirm

**Comment by john-zou** [comment]
> Is there a way to render `JavaArchipelago` in react?

`JavaArchipelago` has position information for each island (`x`, `y`) so May can render islands at the specific positions. It also has information the imports and dependencies so Dara can render the arrows -- I will fix that to make it consistent with the JSON

File: client/src/util/helpers.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (john-zou) [reply client/src/util/helpers.ts:24]: I didn't take multiple classes into consideration at all... I thought Java had 1 class & 1 island per file. I'll make the necessary changes

**Comment by MayKikuchi** [comment]
> > Is there a way to render `JavaArchipelago` in react?
> 
> `JavaArchipelago` has position information for each island (`x`, `y`) so May can render islands at the specific positions. It also has information the imports and dependencies so Dara can render the arrows -- I will fix that to make it consistent with the JSON

does x and y refer to where the top left corner of the island should go, or the center of the island?

**Comment by john-zou** [comment]
> does x and y refer to where the top left corner of the island should go, or the center of the island?

Currently meant for center, but I intend to add top left corner too

File: client/src/components/ImportPanel/index.tsx [hunk 6]
```diff
@@ -93,8 +94,14 @@ const ImportPanel: React.FC<ImportPanelProps> = (props) => {
...
-        alert("Success");
-        setData(res.data);
+        console.log(res.data);
+        console.log("Creating Archipelago");
+        console.time("Archipelago");
+        const javaArchipelago = new JavaArchipelago(res.data);
+        console.timeEnd("Archipelago");
+        console.log("Done!");
+        alert("Success.");
+        setData(javaArchipelago);
```
Comment (algobot76) [inline client/src/components/ImportPanel/index.tsx:104]: I think those `console.log` statements should be removed before merging this PR.

Comment (john-zou) [reply client/src/components/ImportPanel/index.tsx]: Ok, I deleted them

**Pull request merged**